compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirDeclarationsResolveTransformer.kt

```diff
@@ -509,6 +515,42 @@ open class FirDeclarationsResolveTransformer(transformer: FirBodyResolveTransfor
         return result.compose()
     }
 
+    /*
+     * Suppose a function without an explicit return type just returns an anonymous object:
+     *
+     *   fun foo(...) = object : ObjectSuperType {
+     *     override fun ...
+     *   }
+     *
+     * Without unwrapping, the return type ended up with that anonymous object (<no name provided>), while the resolved super type, which
+     * acts like an implementing interface, is a better fit.
+     * In fact, exposing an anonymous object types is prohibited for certain cases, e.g., KT-33917.
+     */
+    private fun FirTypeRef.hideAnonymousObjectIfNeeded(function: FirSimpleFunction?, visibility: Visibility?): FirTypeRef {
+        if (visibility == null) {
+            return this
+        }
+        // Approximate types for non-private (all but package private or private) members.
+        // Also private inline functions, as per KT-33917.
+        if (visibility == Visibilities.Public ||
+            visibility == Visibilities.Protected ||
+            visibility == Visibilities.Internal ||
+            (visibility == Visibilities.Private && function?.isInline == true)
+        ) {
+            val anonymousObjectSupertype =
+                ((((this as? FirResolvedTypeRef)
+                    ?.type as? ConeClassLikeType)
+                    ?.lookupTag as? ConeClassLookupTagWithFixedSymbol)
+                    ?.symbol as? FirAnonymousObjectSymbol)?.fir
```

Comment:
Generally I would say that situation with local class should be similar, so (probably) we should hide type with local visibility if outer callable is public / protected / internal / private inline.

Reply:
Extended to general local types, thanks!

Reply:
Returning a local non-anonymous type from a public function is actually just an error (EXPOSED_FUNCTION_RETURN_TYPE). Anonymous types with multiple supertypes should result in the same, by the way, and it's weird that FirExposedVisibilityDeclarationChecker isn't currently throwing an error on any public function with anonymous object return type.

Reply:
Good to know there is a designated checker for that, and it looks like there are more detailed cases there. I'm not sure either why it isn't currently throwing any errors. Are those checkers invoked/involved in `fir` pipeline?

Regarding multiple supertypes, changed to return an error type instead.